### PR TITLE
Ensure deterministic graph calculation with consistent layer, node, and edge ordering in Kedro-Viz

### LINKED ISSUE #2057: Spike: Try to reproduce non-deterministic rendering
## Description
I'm always frustrated when I do `kedro viz run` twice and each time the rendered DAG is different.

## Context
I was reviewing #1966 and I wanted to see if the rendered pipeline is the same, but it's difficult because the sorting of the nodes can differ.

## Checklist

- [x] Include labels so that we can categorise your feature request

### PR DESCRIPTION
**Pull request by rashidakanchwala**
## Description

Resolves #2057. 

Based on @astrojuanlu's [comment](https://github.com/kedro-org/kedro-viz/issues/2057#issuecomment-2338567627), I realized that Kedro-Viz might be causing the randomness. To investigate, I checked if the inputs to graph calculations—nodes, edges, and layout—were consistent. While the order of nodes and layout remained stable, the order of edges varied with each backend data request. To address this, we now sort the layers, the edges and nodes to ensure consistency.

Update - 
As @ravi-kumar-pilla noted, there was an issue with layer ordering: changing a node name could alter the layer order, especially for layers with identical dependencies, like `model_input` and `feature` in the example he shared. This issue stemmed from the `layer_dependencies` dictionary in `services/layers.py`, where layers with the same dependencies weren’t consistently ordered. To fix this, I added alphabetical sorting for layers with identical dependencies to ensure stability in `toposort`.

For `nodes` and `edges`, I now sort them immediately upon loading from the backend API, ensuring they are consistently ordered in the Redux state.

For testing, I initially considered using Cypress/backend e2e testing with screenshot comparison of the flowchart, but it proved too complex. Instead, I created a new mock dataset, `reordered_spaceflights`, with reordered `nodes` and `edges`. I added tests in `normalise-data-test` and `actions/graph-test`. The first test verifies that nodes and edges are consistently sorted by their ids, regardless of backend order. The second test compares the x and y coordinates in the flowchart, confirming that the graph layout is the same between the two mocks.

## QA notes

- `make build`
- ` pip install -e package`
- `kedro viz --autoreload`

Make changes to the files, doc changes, and see the layout is consistent. 

## Checklist

- [ ] Read the [contributing](/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added new entries to the `RELEASE.md` file
- [ ] Added tests to cover my changes

### CODE REVIEW

**Comment by jitu5** [comment]
@rashidakanchwala I tested with both your branch and main, I made the changes roughly 5 to 7 times and took screenshot every time and compared, it look exactly same on your branch and on main its different every time.

Its working for me.

**Review by jitu5 — APPROVED** [review]
@rashidakanchwala Really cool, Not easy to fix such bugs. 🎉

**Comment by ravi-kumar-pilla** [comment]
Hi @rashidakanchwala , The fix looks promising. I tried without expanding modular pipelines and the nodes seem to position similarly. However, upon expanding modular pipelines, there seems to be some random positioning like below -

**Run 1 -**

image

**Run 2 -**

image

**Observation:** `create_feature_importance` node has changed its position, so are the layers `model_input` and `feature`

**Steps to reproduce:**
1. Follow QA steps
2. Expand all modular pipelines, observe layout of `feature_engineering` 
3. Change node name of `apply_types_to_companies` to `apply_types_to_companies_test` inside `data_ingestion`
4. Observe the layout with the change

**Some questions:**

1. Can we try sorting nodes and layers as well for consistency ? 
2. Is there any overhead or lag you observed due to sorting as `graphNew` is an expensive operation ?
3. Do we think of adding any tests to be sure of the positioning ?

Thank you

File: src/utils/graph/index.js [hunk 1]
```diff
@@ -7,6 +7,9 @@ import { graph } from './graph';
  * which don't affect layout.
  */
 export const graphNew = ({ nodes, edges, layers }) => {
+  // Sort edges to ensure consistent ordering and eliminate randomness in graph calculations
+  const sortedEdges = edges.sort((a, b) => a.id.localeCompare(b.id));
```
Comment (astrojuanlu) [inline src/utils/graph/index.js:11]: ❤️

File: src/actions/graph.test.js [hunk 2]
```diff
@@ -29,19 +41,51 @@ describe('graph actions', () => {
...
+              nodes: expect.any(Array),
+              edges: expect.any(Array),
+              size: expect.any(Object),
+            })
+          );
+        }
+      );
+    });
+
+    it('compares deterministic flowchart of two differently ordered same projects', () => {
```
Comment (ravi-kumar-pilla) [inline src/actions/graph.test.js:60]: Nice !

**Review by ravi-kumar-pilla — APPROVED** [review]
This is really good @rashidakanchwala , tested few times and the order seems the same. Great work !! 💯

**Review by jitu5 — APPROVED** [review]
Thanks @rashidakanchwala, I tested again. Working as expected.

**Review by Huongg — APPROVED** [review]
Thanks @rashidakanchwala, the changes look good to me, and everything is working as expected locally for me too

**Pull request merged**